image/image_linux.go

```diff
@@ -272,302 +273,228 @@ func makeLabel(now time.Time) string {
-			return fmt.Errorf("cannot prepare seed over existing system or an already booted image, detected state file %s", dirs.SnapStateFileUnder(rootDir))
-		}
-		if snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDirUnder(rootDir), "*.snap")); len(snaps) > 0 {
-			return fmt.Errorf("expected empty snap dir in rootdir, got: %v", snaps)
-		}
+	w     *seedwriter.Writer
+	model *asserts.Model
+	tsto  *tooling.ToolingStore
+}
 
+func (is *imageSeeder) modelessDirs(opts *Options) error {
+	if opts.Classic {
+		// Classic, PrepareDir is the root dir itself
+		is.rootDir = opts.PrepareDir
 	} else {
-		// Core 20, writing for the system-seed partition
-		seedDir = filepath.Join(opts.PrepareDir, "system-seed")
-		label = makeLabel(time.Now())
-		bootRootDir = seedDir
-
-		// validity check target
-		if systems, _ := filepath.Glob(filepath.Join(seedDir, "systems", "*")); len(systems) > 0 {
-			return fmt.Errorf("expected empty systems dir in system-seed, got: %v", systems)
-		}
+		// Core 16/18,  writing for the writeable partition
+		is.rootDir = filepath.Join(opts.PrepareDir, "image")
+		is.bootRootDir = is.rootDir
 	}
+	is.seedDir = dirs.SnapSeedDirUnder(is.rootDir)
 
-	// TODO: developer database in home or use snapd (but need
-	// a bit more API there, potential issues when crossing stores/series)
-	db, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
-		Backstore: asserts.NewMemoryBackstore(),
-		Trusted:   trusted,
-	})
-	if err != nil {
-		return err
+	// validity check target
+	if osutil.FileExists(dirs.SnapStateFileUnder(is.rootDir)) {
+		return fmt.Errorf("cannot prepare seed over existing system or an already booted image, detected state file %s", dirs.SnapStateFileUnder(is.rootDir))
+	}
+	if snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDirUnder(is.rootDir), "*.snap")); len(snaps) > 0 {
+		return fmt.Errorf("expected empty snap dir in rootdir, got: %v", snaps)
 	}
+	return nil
+}
 
-	wOpts := &seedwriter.Options{
-		SeedDir:        seedDir,
-		Label:          label,
-		DefaultChannel: opts.Channel,
+func (is *imageSeeder) modeDirs(opts *Options) error {
+	// Core 20, writing for the system-seed partition
+	is.seedDir = filepath.Join(opts.PrepareDir, "system-seed")
+	is.label = makeLabel(time.Now())
+	is.bootRootDir = is.seedDir
 
-		TestSkipCopyUnverifiedModel: osutil.GetenvBool("UBUNTU_IMAGE_SKIP_COPY_UNVERIFIED_MODEL"),
+	// validity check target
+	if systems, _ := filepath.Glob(filepath.Join(is.seedDir, "systems", "*")); len(systems) > 0 {
+		return fmt.Errorf("expected empty systems dir in system-seed, got: %v", systems)
 	}
+	return nil
+}
 
-	w, err := seedwriter.New(model, wOpts)
-	if err != nil {
-		return err
+func makeImageSeeder(tsto *tooling.ToolingStore, model *asserts.Model, opts *Options) (*imageSeeder, error) {
+	// Determine image seed paths, which can vary based on the type of image
+	// we are generating.
+	is := &imageSeeder{
+		hasModes: model.Grade() != asserts.ModelGradeUnset,
```

Comment:
we can probably also stick opts in here?

Reply:
Instead of just storing a copy of opts, I have updated the code to extract which members we need from the `Options` structure and store a copy inside `imageSeeder`, as we don't need a whole lot from it, and it's nice to see what members are actually needed.